### PR TITLE
[R-package] fix segfaults caused by missing Booster and Dataset handles (fixes #4208)

### LINKED ISSUE #4208: [R-package] R handles produce segmentation faults when de-serialized
The R interface uses a "handle" object which I guess stores a pointer to a C++ object. The objects these handles point to do not survive serialization in R, and will cause segmentation faults if they are serialized and de-serialized - which happens for example when restarting an R session.

Example:
```r
library(lightgbm)
data(agaricus.train, package = "lightgbm")
train <- agaricus.train
dtrain <- lgb.Dataset(train$data, label = train$label)
data(agaricus.test, package = "lightgbm")
test <- agaricus.test
dtest <- lgb.Dataset.create.valid(dtrain, test$data, label = test$label)
params <- list(objective = "regression", metric = "l2")
valids <- list(test = dtest)
model <- lgb.train(
    params = params
    , data = dtrain
    , nrounds = 5L
    , valids = valids
    , min_data = 1L
    , learning_rate = 1.0
    , early_stopping_rounds = 3L
)
```
After running it, restart the R session (assuming it's configured to save the environment between restarts), and execute these two lines again:
```r
library(lightgbm)
model <- lgb.train(
    params = params
    , data = dtrain
    , nrounds = 5L
    , valids = valids
    , min_data = 1L
    , learning_rate = 1.0
    , early_stopping_rounds = 3L
)
```
At that point it will produce a segmentation fault, crashing the R process.

I suppose the correct solution would be to use R's own external pointer object class and leave the destructor/free-er to the same R external pointer object. Those objects will also reset to `nullptr` upon de-serialization, which can then be checked inside the functions beforehand to avoid producing segmentation faults.

### PR DESCRIPTION
**Pull request by jameslamb**
Fixes #4208, as part of #4310.

Currently, the R package can segfault if Booster or Dataset methods are called on an object where the handle to the C++ side is null. This situation can happen if, for example, someone uses `saveRDS()` on one of these objects and then loads it with `readRDS()`. That is exactly what happens when restarting an R session, if you have R configured to save your workspace and reload it automatically.

This PR proposes changes to ensure that an informative error is raised in such situations, instead of a segfault.

### Notes for Reviewers

#4296 documents the desire for `{lightgbm}` to guarantee that a Booster / Dataset saved with `saveRDS()` is *usable* after loading it with `readRDS()`. That might be addressed in a future release, but it's out of scope for release 3.3.0 and this PR. At least as of this PR, users will experience informative errors instead of their R sessions crashing.

I was not able to trigger the `_AssertDatasetHandleNotNull()` error message in tests, since all paths user-facing paths through the `Dataset` object either already have protective checks using `lgb.is.null.handle()` or raise other errors before even trying to invoke C++ code. But I still think those asserts are worth adding, to catch errors that might be introduced in future refactorings of the R package or in other code paths that are missed in tests.

### CODE REVIEW

**Comment by jameslamb** [comment]
valgrind just timed out after 3 hours, trying again 😫 

https://github.com/microsoft/LightGBM/runs/3500789328?check_suite_focus=true

**Comment by jameslamb** [comment]
I see this in one place in the test logs for the valgrind job: https://github.com/microsoft/LightGBM/runs/3501825714?check_suite_focus=true

```text
......................Error in self$construct() : 
  Attempting to create a Dataset without any raw data. This can happen if you have called Dataset$finalize() or if this Dataset was saved with saveRDS(). To avoid this error in the future, use lgb.Dataset.save() or Dataset$save_binary() to save lightgbm Datasets.
```

Hard to tell exactly which test, but I see `saveRDS.lgb.Booster() and readRDS.lgb.Booster():` shortly after, so it must be between these lines.

https://github.com/microsoft/LightGBM/blob/417ba1921797d3dc5787ae8d297dcbbfc7bd7184/R-package/tests/testthat/test_lgb.Booster.R#L640

https://github.com/microsoft/LightGBM/blob/417ba1921797d3dc5787ae8d297dcbbfc7bd7184/R-package/tests/testthat/test_lgb.Booster.R#L881

Interestingly, I don't see that same error in the R 4.0 Windows MSVC job.

https://github.com/microsoft/LightGBM/pull/4586/checks?check_run_id=3500712973

And I'm also curious why that is showing up in the logs but the `valgrind` test didn't actually fail.

Will have to investigate it tomorrow.

**Comment by jameslamb** [comment]
Ah ok, figured it out! The error in the logs of https://github.com/microsoft/LightGBM/pull/4586#issuecomment-912231412 is not an issue.

It's coming from this new test added in this PR:

```r
test_that("Booster$new() using a Dataset with a null handle should raise an informative error and not segfault", {
    data(agaricus.train, package = "lightgbm")
    train <- agaricus.train
    dtrain <- lgb.Dataset(train$data, label = train$label)
    dtrain$construct()
    tmp_file <- tempfile(fileext = ".bin")
    saveRDS(dtrain, tmp_file)
    rm(dtrain)
    dtrain <- readRDS(tmp_file)
    expect_error({
        bst <- Booster$new(train_set = dtrain)
    }, regexp = "lgb.Booster: cannot create Booster handle")
})
```

That "error" is intentionally being caught in this try-catch:

https://github.com/microsoft/LightGBM/blob/b7120d27db6d5a7c3c3fe6b0a823552bfe2b4a85/R-package/R/lgb.Booster.R#L33

https://github.com/microsoft/LightGBM/blob/b7120d27db6d5a7c3c3fe6b0a823552bfe2b4a85/R-package/R/lgb.Booster.R#L113

So when users actually run code similar to that test, they'll get the error "cannot create Booster handle" but then also see the printed text about "Attempting to create a Dataset...".

image

> Interestingly, I don't see that same error in the R 4.0 Windows MSVC job.

This also is not a problem. It's probably because of how we have to redirect stderr in Windows CI jobs.

https://github.com/microsoft/LightGBM/blob/b7120d27db6d5a7c3c3fe6b0a823552bfe2b4a85/.ci/test_r_package_windows.ps1#L23

**Review by StrikerRUS — COMMENTED** [review]
This PR has a lot of potential merge conflicts with #4597 PR. Which one would you like to merge first?

**Comment by jameslamb** [comment]
> This PR has a lot of potential merge conflicts with #4597 PR. Which one would you like to merge first?

Let's merge #4597 first. I'd rather take on the pain of merge conflicts than inflict them on a non-maintainer 😂

File: R-package/src/lightgbm_R.cpp
Comment (jameslamb) [reply R-package/src/lightgbm_R.cpp:449]: oh interesting! I was so focus on "add the booster check to the Booster methods and Dataset check to Dataset methods", I missed this one. I'll have to check the implementation in `c_api.cpp` to see if it can handle a null Dataset handle (my guess is no, but need to confirm)

File: R-package/tests/testthat/test_dataset.R
Comment (jameslamb) [reply R-package/tests/testthat/test_dataset.R:536]: ah! thank you for catching this

**Comment by jameslamb** [comment]
I'll wait to update this until #4613 is also merged.

File: R-package/src/lightgbm_R.cpp
Comment (shiyu1994) [reply R-package/src/lightgbm_R.cpp:488]: This method is for adding valid data. According to the implementation, it will takes the valid_data as non-null.

File: R-package/src/lightgbm_R.cpp
Comment (shiyu1994) [reply R-package/src/lightgbm_R.cpp:497]: Same as LGBM_BoosterAddValidData_R

File: R-package/src/lightgbm_R.cpp
Comment (jameslamb) [reply R-package/src/lightgbm_R.cpp:488]: perfect, thanks for the help @shiyu1994 !

File: R-package/src/lightgbm_R.cpp
Comment (jameslamb) [inline R-package/src/lightgbm_R.cpp:558]:
```suggestion
  _AssertBoosterHandleNotNull(handle);
  R_API_BEGIN();
```
```suggestion
  R_API_BEGIN();
  _AssertBoosterHandleNotNull(handle);
```

**Review by StrikerRUS — APPROVED** [review]
LGTM! 👍

**Pull request merged**
